proto/lambda.proto

```diff
@@ -0,0 +1,195 @@
+message Type {
+    message Unit { }
+    message Int { }
+    message Int32 { }
+    message Int64 { }
+    message Bool { }
+    message String { }
+    message Lwt { }
+    message List {
+        required Type value = 1;
+    }
+    message Array {
+        required Type value = 1;
+    }
+    message Option {
+        required Type value = 1;
+    }
+    message Apply {
+        required Type a = 1;
+        required Type b = 2;
+    }
+    message Arrow {
+        required Type a = 1;
+        required Type b = 2;
+    }
+    message Pair {
+        required Type a = 1;
+        required Type b = 2;
+    }
+    message Either {
+        required Type a = 1;
+        required Type b = 2;
+    }
+    message Result {
+        required Type a = 1;
+        required Type b = 2;
+    }
+    message Abstract {
+        required string witness = 1;
+    }
+
+    oneof t {
+        Unit unit = 1;
+        Int int = 2;
+        Int32 int32 = 3;
+        Int64 int64 = 4;
+        Bool bool = 5;
+        String string = 6;
+        Lwt lwt = 7;
+        List list = 8;
+        Array array = 9;
+        Option option = 10;
+        Apply apply = 11;
+        Arrow arrow = 12;
+        Pair pair = 13;
+        Either either = 14;
+        Result result = 15;
+        Abstract abstract = 16;
+    }
+}
+
+message Primitive {
+    required string name = 1;
+    repeated Type arguments = 2;
+    required Type return = 3;
+    required int32 smartptr = 4;
```

Comment:
What value should the Haskell side provide for the `smartptr` field? Maybe this field should not be serialized, but resolved by the Mirage server based on the `name` field?

Reply:
The idea behind `smartptr` is to bind (on the MirageOS/OCaml side) an `int64` with functions/primitives. Then, when we deserialize, we try to find this identifier in an internal map of the unikernel. So, currently the identifier is an `int64` to be close to the idea of _how to share a function between two processes_. However, we can choose another identifier (like a `string` if you want).

On the haskell side, you just need to have a `Map int64 ([value] -> value)` (or, if you prefer a `string`, a `Map String ([value] -> value)`.

Reply:
Exactly, so I will change the key to be a `string` 👍